code/go/0chain.net/blobbercore/filestore/storage.go

```diff
@@ -0,0 +1,638 @@
+	}
+	defer f.Close()
+
+	fStat, err := f.Stat()
+	if err != nil {
+		return false, common.NewError("stat_error", err.Error())
+	}
+
+	fileSize := fStat.Size()
+
+	var hash string
+	switch {
+	case fileData.IsThumbnail:
+		h := sha256.New()
+		_, err := io.Copy(h, f)
+		if err != nil {
+			return false, common.NewError("read_error", err.Error())
+		}
+		hash = hex.EncodeToString(h.Sum(nil))
+	default:
+		/* Uncomment it after padding is done in gosdk
+		if fileSize > fileData.ChunkSize && fileSize%fileData.ChunkSize != 0 { // workaround for data without padding
+			return false, common.NewError("invalid_data",
+				fmt.Sprintf("file size %d is not exactly divisible by chunk size %d", fileSize, fileData.ChunkSize))
+		}
+		*/
+
+		hasher := sdk.CreateHasher(int(fileData.ChunkSize))
+
+		n := int64(math.Ceil(float64(fileSize) / float64(fileData.ChunkSize)))       // workaround for data without padding otherwise fileSize/fileData.ChunkSize
+		n = int64(math.Max(float64(1), float64(n)))                                  // workaround for data without padding otherwise non-existing line
+		chunkSize := int64(math.Min(float64(fileSize), float64(fileData.ChunkSize))) // workaround for data without padding otherwise non-existing line
+
+		for i := int64(0); i < n; i++ {
+			offset := i * chunkSize
+			data := make([]byte, chunkSize)
+			n, err := f.ReadAt(data, offset)
+			if err != nil && !errors.Is(err, io.EOF) {
+				return false, common.NewError("read_error", err.Error())
+			}
+
+			/* Uncomment when padding is done in gosdk
+			if n != int(chunkSize) {
+				return false, common.NewError("read_error",
+					fmt.Sprintf("expected read %d, got %d", chunkSize, n))
+			}
+			*/
+
+			h := sha256.New()
+			_, err = h.Write(data[:n]) // workaround for data without padding otherwise h.Write(data)
+			if err != nil {
+				return false, common.NewError("hash_write_error", err.Error())
+			}
+
+			err = hasher.WriteHashToContent(hex.EncodeToString(h.Sum(nil)), int(i))
+			if err != nil {
+				return false, common.NewError("content_hash_write_error", err.Error())
+			}
+		}
+
+		hash, err = hasher.GetContentHash()
+		if err != nil {
+			return false, common.NewError("get_content_hash_error", err.Error())
+		}
+	}
+
+	if hash != fileData.Hash {
+		return false, common.NewError("invalid_hash", "calculated content hash does not match with client's content hash")
+	}
+
+	key := getKey(allocID, fileData.Hash)
+	l, _ := contentHashMapLock.GetLock(key)
+	l.Lock()
+	defer func() {
+		if err != nil {
+			l.Unlock()
+		}
+	}()
+
+	fileObjectPath, err := fs.GetPathForFile(allocID, fileData.Hash)
+	if err != nil {
+		return false, common.NewError("get_file_path_error", err.Error())
+	}
+
+	err = createDirs(filepath.Dir(fileObjectPath))
+	if err != nil {
+		return false, common.NewError("blob_object_dir_creation_error", err.Error())
+	}
+
+	//move file from tmp location to the objects folder
+	err = os.Rename(tempFilePath, fileObjectPath)
+	l.Unlock()
```

Comment:
The l.Unlock() will be called twice if the os.Rename() return err, as the defer function above will call it again.